File: Sem3/Advanced-Programming/Python/queue.py
class Queue:
    def __init__(self):
        self.queue = []

    def enqueue(self, item):
        self.queue.append(item)

    def dequeue(self):
        if len(self.queue) == 0:
            raise IndexError("Dequeue from an empty queue")
        return self.queue.pop(0)
    
    def front(self):
        if len(self.queue) == 0:
            raise IndexError("Peek from an empty queue")
        return self.queue[0]
    
    def is_empty(self):
        return len(self.queue) == 0
    
    def size(self):
        return len(self.queue)
    
q = Queue()
q.enqueue(15)
q.enqueue(25)
q.enqueue(50)
print(q.is_empty())
print(q.size())
print(q.front())
q.dequeue()
q.dequeue()
print(q.size())
print(q.front())

File: Sem4/DSA/Programs/Queue.c
#include <stdio.h>
#define MAX 5

int queue[MAX];
int front = -1;
int rear = -1;

void enqueue(int data) {
    if (rear == MAX - 1) {
        printf("Queue Overflow\n");
        return;
    }
    if (front == -1) {
        front = 0;
    }
    rear++;
    queue[rear] = data;
}

int dequeue() {
    if (front == -1 || front > rear) {
        printf("Queue Underflow\n");
        return -1;
    }
    int data = queue[front];
    front++;
    if (front > rear) {
        front = rear = -1;
    }
    return data;
}

void display() {
    if (front == -1 || front > rear) {
        printf("Queue is empty\n");
        return;
    }
    printf("Queue elements: ");
    for (int i = front; i <= rear; i++) {
        printf("%d ", queue[i]);
    }
    printf("\n");
}

int main() {
    enqueue(10);
    enqueue(20);
    enqueue(30);
    enqueue(40);

    display();

    dequeue();
    dequeue();
    dequeue();
    dequeue();
    
    display();
    
    return 0;

}
